google_guest_agent/agentcrypto/mtls_mds.go

```diff
@@ -208,13 +217,119 @@ func (j *CredsJob) Interval() (time.Duration, bool) {
 	return MTLSScheduleInterval, true
 }
 
-// ShouldEnable returns true if MDS endpoint for fetching credentials is available on the VM.
-// Used for identifying if we want schedule bootstrapping and enable MDS mTLS credential rotation.
+// ShouldEnable implements scheduler job interface which returns true if job
+// should be scheduled based on previous cached [isEnabled] value.
 func (j *CredsJob) ShouldEnable(ctx context.Context) bool {
+	return j.isEnabled.Load()
+}
+
+// checkUserSettings checks and stores user settings for job enablement and the use
+// of OS Native store.
+func (j *CredsJob) checkUserSettings(ctx context.Context, mds *metadata.Descriptor) {
+	logger.Debugf("Checking user settings for %s", j.ID())
+	j.useNativeStore.Store(j.shouldUseNativeStore(mds))
+	j.isEnabled.Store(j.shouldEnableJob(ctx, mds))
+}
+
+// shouldEnableJob returns true if MDS endpoint for fetching credentials is available on the VM
+// and user has not disabled the bootstrapping via config file or Metadata.
+// Used for identifying if we want schedule bootstrapping and enable MDS mTLS credential rotation.
+func (j *CredsJob) shouldEnableJob(ctx context.Context, mds *metadata.Descriptor) bool {
+	var enable bool
+
+	if cfg.Get().MDS != nil {
+		enable = !cfg.Get().MDS.DisableHTTPSMdsSetup
+		logger.Debugf("Found instance config file attribute for enable credential refrehser set to: %t", enable)
+	}
+
+	if mds.Project.Attributes.DisableHTTPSMdsSetup != nil {
+		enable = !*mds.Project.Attributes.DisableHTTPSMdsSetup
+		logger.Debugf("Found project level attribute for enable credential refrehser set to: %t", enable)
+	}
+
+	if mds.Instance.Attributes.DisableHTTPSMdsSetup != nil {
+		enable = !*mds.Instance.Attributes.DisableHTTPSMdsSetup
+		logger.Debugf("Found instance level attribute for enable credential refrehser set to: %t", enable)
+	}
+
 	_, err := j.client.GetKey(ctx, clientCertsKey, nil)
 	if err != nil {
-		logger.Warningf("Skipping scheduling credential generation job, failed to reach client credentials endpoint(%s) with error: %v", clientCertsKey, err)
+		logger.Debugf("Skipping scheduling credential generation job, failed to reach client credentials endpoint(%s) with error: %v", clientCertsKey, err)
+		enable = false
+	}
+
+	return enable
+}
+
+// shouldUseNativeStore checks if user has configured agent to use OS Native Store for
+// storing credentials.
+func (j *CredsJob) shouldUseNativeStore(mds *metadata.Descriptor) bool {
+	var useNative bool
+
+	if mds.Instance.Attributes.HTTPSMDSSkipNativeStore != nil {
```

Comment:
the project vs instance is inverted as well, it should follow the `shouldEnableJob()` implementation, you got it correct there.

Reply:
isn't that same? `shouldEnableJob` is just storing as we also want to check mds endpoint exists but here we're simply returning if instance is set to true/false it doesn't matter what project is as it will always take higher precedence.

Reply:
Can we follow the same approach for both and avoid this confusion again? :-)

Reply:
Done